Src/FluentAssertions/StringContainmentConstraints.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Diagnostics;
+using FluentAssertions.Primitives;
+
+namespace FluentAssertions
+{
+    [DebuggerNonUserCode]
+    public class StringContainmentConstraints : AndConstraint<StringAssertions>
+    {
+        private readonly string expected;
+        private readonly int occurrences;
+        private readonly StringAssertions parentConstraint;
+
+        public StringContainmentConstraints(StringAssertions parentConstraint, string expected) : base(parentConstraint)
+        {
+            this.parentConstraint = parentConstraint;
+            this.expected = expected;
+            occurrences = CountOccurrences(parentConstraint.Subject, expected, StringComparison.Ordinal);
+        }
+
+        private static int CountOccurrences(string actual, string expected, StringComparison comparison)
+        {
+            actual = actual ?? "";
+            expected = expected ?? "";
+
+            int occurrences = 0;
+
+            int index;
+
+            while ((index = actual.IndexOf(expected, comparison)) >= 0)
```

Comment:
x Use [`public int IndexOf (string value, int startIndex, int count, StringComparison comparisonType)`](https://docs.microsoft.com/en-us/dotnet/api/system.string.indexof?view=netframework-4.7.2#System_String_IndexOf_System_String_System_Int32_System_Int32_System_StringComparison_) to avoid implementing a variant of [Schlemiel the Painter's Algorithm](https://en.wikichip.org/wiki/schlemiel_the_painter%27s_algorithm). This will also avoid any calls Substring.
Here's an [implementation](https://www.dotnetperls.com/string-occurrence)

Reply:
Good point. I will fix it